### PR TITLE
BTree Index Compression

### PR DESCRIPTION
**Pull request by linhong-intel**
## What changes were proposed in this pull request?

Compress / Decompress BTree Index file

## How was this patch tested?

current UT

### CODE REVIEW

**Comment by linhong-intel** [comment]
Actual changes included in commit: https://github.com/Intel-bigdata/OAP/pull/709/commits/7c23a04a24b346b16354250e223fa4ff0abb41f2

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/ColumnValues.scala [hunk 1]
```diff
@@ -20,14 +20,17 @@ package org.apache.spark.sql.execution.datasources.oap
...
 import org.apache.spark.sql.catalyst.util.{ArrayData, MapData}
-import org.apache.spark.sql.execution.datasources.oap.filecache.FiberCache
+import org.apache.spark.sql.execution.datasources.oap.filecache.{FiberCache, FiberCacheReleasable}
 import org.apache.spark.sql.types._
 import org.apache.spark.unsafe.types.{CalendarInterval, UTF8String}
 import org.apache.spark.util.collection.BitSet
 
-class ColumnValues(defaultSize: Int, dataType: DataType, val buffer: FiberCache) {
+class ColumnValues(defaultSize: Int, dataType: DataType, val buffer: FiberCache)
+    extends FiberCacheReleasable {
```
Comment (adrian-wang) [inline src/main/scala/org/apache/spark/sql/execution/datasources/oap/ColumnValues.scala]: Does this pr have overlap with #696 ?

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/BTreeIndexRecordReaderV2.scala [hunk 2]
```diff
@@ -0,0 +1,420 @@
...
+    }
+  }
+
+  /**
+   * Compare auxiliary function.
+   * @param x, y are the key to be compared.
+   *         One comes from interval.start/end. One comes from index records
+   * @param isStart indicates to compare interval.start or interval end
+   */
+  private[index] def rowOrdering(x: InternalRow, y: InternalRow, isStart: Boolean): Int = {
```
Comment (adrian-wang) [inline src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/BTreeIndexRecordReaderV2.scala]: should we extract some common methods to a common base class for different reader versions?

File: src/main/scala/org/apache/spark/sql/internal/oap/OapConf.scala [hunk 3]
```diff
@@ -202,7 +202,7 @@ object OapConf {
       .internal()
       .doc("The writer version of BTree index")
       .stringConf
-      .createWithDefault("v1")
+      .createWithDefault("v2")
```
Comment (adrian-wang) [inline src/main/scala/org/apache/spark/sql/internal/oap/OapConf.scala]: add `.checkValues(Set("v1", "v2"))` to this?

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/IndexUtils.scala [hunk 4]
```diff
@@ -205,4 +206,21 @@ private[oap] object IndexUtils {
       (m, found)
     }
   }
+
+  private val CODEC_MAGIC: Array[Byte] = "CODEC".getBytes("UTF-8")
+
+  def compressIndexData(compressor: BytesCompressor, bytes: Array[Byte]): Array[Byte] = {
+    CODEC_MAGIC ++ toBytes(bytes.length) ++ compressor.compress(bytes)
```
Comment (adrian-wang) [inline src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/IndexUtils.scala]: The nodes here are compressed by each own. Shall we try run length/delta encoding across tree nodes, and then compress one loading unit together, maybe in v3?

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/ColumnValues.scala (hunk 1 above)
Comment (linhong-intel) [reply src/main/scala/org/apache/spark/sql/execution/datasources/oap/ColumnValues.scala]: I'll rebase this after #696

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/IndexUtils.scala (hunk 4 above)
Comment (linhong-intel) [reply src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/IndexUtils.scala]: sure.

**Comment by linhong-intel** [comment]
@adrian-wang 
updated PR, please review this again.

**Comment by linhong-intel** [comment]
Will rebase this on #711

**Comment by linhong-intel** [comment]
@adrian-wang ready to review.
Thanks.

File: src/test/scala/org/apache/spark/sql/execution/datasources/oap/index/BTreeIndexScannerSuite.scala [hunk 5]
```diff
@@ -133,7 +132,7 @@ class BTreeIndexScannerSuite extends SharedOapContext {
     val schema = StructType(StructField("col", IntegerType) :: Nil)
     val path = new Path(Utils.createTempDir().getAbsolutePath, "tempIndexFile")
     val fileWriter = IndexFileWriterImpl(configuration, path)
-    val writer = BTreeIndexRecordWriter(configuration, fileWriter, schema)
+    val writer = BTreeIndexRecordWriterV1(configuration, fileWriter, schema)
```
Comment (adrian-wang) [inline src/test/scala/org/apache/spark/sql/execution/datasources/oap/index/BTreeIndexScannerSuite.scala]: why test v1 here?

File: src/test/scala/org/apache/spark/sql/execution/datasources/oap/FilterSuite.scala [hunk 6]
```diff
@@ -134,8 +134,14 @@ class FilterSuite extends QueryTest with SharedOapContext with BeforeAndAfterEac
...
   test("filtering") {
-    val data: Seq[(Int, String)] = (1 to 300).map { i => (i, s"this is test $i") }
-    data.toDF("key", "value").createOrReplaceTempView("t")
+    val rowRDD = spark.sparkContext.parallelize(1 to 300, 2).map { i =>
+      Seq(i, s"this is test $i")
+    }.map(Row.fromSeq)
+    val schema =
+      StructType(StructField("a", IntegerType) :: StructField("c", StringType) :: Nil)
+    val df = spark.createDataFrame(rowRDD, schema)
+    df.createOrReplaceTempView("t")
```
Comment (adrian-wang) [inline src/test/scala/org/apache/spark/sql/execution/datasources/oap/FilterSuite.scala]: why we change this part of code? shall we apply the change to all these code snippets in other test cases, in a separate pr?

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/OapIndexOutputFormat.scala [hunk 7]
```diff
@@ -31,6 +32,15 @@ import org.apache.spark.sql.types.StructType
 private[index] class OapIndexOutputFormat extends FileOutputFormat[Void, InternalRow] {
 
   private val BTREE_WRITER_VERSION = OapConf.OAP_INDEX_BTREE_WRITER_VERSION.key
+
+  private def getCodec(taskAttemptContext: TaskAttemptContext): CompressionCodec = {
+    val configuration = ContextUtil.getConfiguration(taskAttemptContext)
+    CompressionCodec.valueOf(
+      configuration.get(
+        OapConf.OAP_INDEX_BTREE_COMPRESSION.key,
```
Comment (adrian-wang) [inline src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/OapIndexOutputFormat.scala]: I can see there are several things related to BTree index only, we can further split `OapIndexOutputFormat` into 2 corresponding formats and do a re-factorize.

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/BTreeIndexRecordWriter.scala [hunk 8]
```diff
@@ -141,28 +158,52 @@ private[index] case class BTreeIndexRecordWriter(
         } else {
           val nodeUniqueKeys = sortedIter.take(keyCount).toArray
           val bTreeNodeMetaData =
-            serializeNode(nodeUniqueKeys, startPosInRowList, tempIdWriter)
+            serializeNode(nodeUniqueKeys, startPosInRowList, tempIdWriter, rowIdListBuffer)
```
Comment (adrian-wang) [inline src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/BTreeIndexRecordWriter.scala]: The buffer seems is only used as a buffer to hold data until we compress them and write to file. So we'd better put the buffer initialization inside of `serializeNode`, right?

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/BTreeIndexRecordWriter.scala [hunk 9]
```diff
@@ -141,28 +158,52 @@ private[index] case class BTreeIndexRecordWriter(
...
       Seq(BTreeNodeMetaData(0, 0, null, null))
     }
-    tempIdWriter.close()
-    // Write Row Id List
-    fileWriter.writeRowId(tempIdWriter)
     var nullIdSize = 0
     nullBitSet.iterator.foreach { x =>
-      fileWriter.write(IndexUtils.toBytes(x))
+      // fileWriter.write(IndexUtils.toBytes(x))
+      storeRowId(x, rowIdListBuffer, tempIdWriter)
```
Comment (adrian-wang) [inline src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/BTreeIndexRecordWriter.scala]: we can also extract this part of writing null row id list to another function.

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/BTreeIndexRecordWriterV2.scala [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+
+import org.apache.spark.sql.execution.datasources.oap.index.OapIndexProperties.IndexVersion
+import org.apache.spark.sql.execution.datasources.oap.io.CodecFactory
+import org.apache.spark.sql.types.StructType
+
+private[index] case class BTreeIndexRecordWriterV2(
+    configuration: Configuration,
+    fileWriter: IndexFileWriter,
+    keySchema: StructType,
+    codec: CompressionCodec = CompressionCodec.UNCOMPRESSED)
```
Comment (adrian-wang) [inline src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/BTreeIndexRecordWriterV2.scala]: A quick question: when `codec` set to `UNCOMPRESSED`, is the output file identical with v1?

File: src/test/scala/org/apache/spark/sql/execution/datasources/oap/index/BTreeIndexScannerSuite.scala (hunk 5 above)
Comment (linhong-intel) [reply src/test/scala/org/apache/spark/sql/execution/datasources/oap/index/BTreeIndexScannerSuite.scala]: my mistake

File: src/test/scala/org/apache/spark/sql/execution/datasources/oap/FilterSuite.scala (hunk 6 above)
Comment (linhong-intel) [reply src/test/scala/org/apache/spark/sql/execution/datasources/oap/FilterSuite.scala]: sure. I use this test case a lot. This change is to control parallelism. I'll revert this.

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/BTreeIndexRecordWriter.scala (hunk 8 above)
Comment (linhong-intel) [reply src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/BTreeIndexRecordWriter.scala]: here is the problem about your idea:
rowIdListPartSize = 64.
Node#1 has 65 rows
If I put the initialization inside `serializeNode `. the extra 1 line has to be return back and pass to next round `serializeNode`.

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/BTreeIndexRecordWriterV2.scala (hunk 10 above)
Comment (linhong-intel) [reply src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/BTreeIndexRecordWriterV2.scala]: no.
every compressed section has a codec value at begin. For example
Data: 1234567
v1: 1234567
v2: CODECx1234567

What's more:
Footer layout is changed to handle compressed row id list partition length is not fixed.

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/BTreeIndexRecordWriter.scala (hunk 9 above)
Comment (linhong-intel) [reply src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/BTreeIndexRecordWriter.scala]: not get you point? more hint?

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/BTreeIndexRecordWriter.scala (hunk 8 above)
Comment (adrian-wang) [reply src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/BTreeIndexRecordWriter.scala]: OK, I understand.

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/BTreeIndexRecordWriter.scala (hunk 9 above)
Comment (adrian-wang) [reply src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/BTreeIndexRecordWriter.scala]: This is related to the misunderstanding above, so just ignore it. Sorry.

**Pull request merged**